File: sdk/typescript/src/response_metadata.ts
/** Advisory API metadata. Codes and levels remain open to future values. */
export interface ResponseNotice {
  code: string;
  level: string;
  message: string;
}

export interface ResponseMetadata {
  notices?: ResponseNotice[];
}

export interface APIResponse<T> extends ResponseMetadata {
  data: T extends void ? null : T;
}

export type ResponseObserver = (metadata: ResponseMetadata) => void | Promise<void>;

/** @internal */
export function parseResponseNotices(value: unknown): ResponseNotice[] | undefined {
  if (!Array.isArray(value)) return undefined;
  const notices = value.filter((item): item is ResponseNotice =>
    item !== null && typeof item === "object"
    && typeof item.code === "string" && typeof item.level === "string" && typeof item.message === "string",
  ).map(({ code, level, message }) => ({ code, level, message }));
  return notices.length ? notices : undefined;
}

/** @internal */
export function collectResponseNotices(target: ResponseNotice[], metadata: ResponseMetadata): void {
  for (const notice of metadata.notices ?? []) {
    if (!target.some((item) => item.code === notice.code && item.level === notice.level && item.message === notice.message)) {
      target.push({ ...notice });
    }
  }
}

/** @internal Optional metadata must never change the primary response outcome. */
export async function observeResponse(
  response: Response,
  url: string,
  requestMethod: string | false,
  ...observers: (ResponseObserver | undefined)[]
): Promise<void> {
  if (!observers.some(Boolean)) return;
  let notices: ResponseNotice[] | undefined;
  let validHeader = false;
  try {
    const header = response.headers?.get("Inkbox-Notices");
    if (header !== null && header !== undefined) {
      const value: unknown = JSON.parse(header);
      notices = parseResponseNotices(value);
      validHeader = value === null || (Array.isArray(value) && (value.length === 0 || notices !== undefined));
    }
  } catch { /* Ignore malformed optional metadata. */ }
  // Only these response contracts declare a top-level metadata field.
  try {
    const path = requestMethod ? new URL(url).pathname.replace(/\/+$/, "") : "";
    const bodyMetadata = /^\/api\/v1\/identities\/[^/]+$/.test(path)
      || (requestMethod === "POST" && path === "/api/v1/identities")
      || (requestMethod === "PUT" && /^\/api\/v1\/identities\/[^/]+\/avatar$/.test(path))
      || /^\/api\/v1\/identities\/[^/]+\/contacts\/[^/]+\/(access|permissions)$/.test(path)
      || /^\/api\/v1\/identities\/[^/]+\/(contact-permissions|contact-communication-policies)$/.test(path)
      || /^\/api\/v1\/contacts\/[^/]+\/(communication-policy|communication-preview)$/.test(path)
      || /^\/api\/v1\/(mail\/mailboxes|phone\/numbers)\/[^/]+$/.test(path);
    if (!validHeader && bodyMetadata && response.ok && typeof response.clone === "function"
      && response.headers?.get("content-type")?.includes("application/json")) {
      const body = await response.clone().json();
      notices = parseResponseNotices(body?.notices);
    }
  } catch { /* Preserve primary payload parsing and errors. */ }
  await notifyResponseObservers(notices ? { notices } : {}, ...observers);
}

/** @internal */
export async function notifyResponseObservers(metadata: ResponseMetadata, ...observers: (ResponseObserver | undefined)[]): Promise<void> {
  for (const observer of observers) {
    if (!observer) continue;
    try {
      await observer(metadata.notices ? { notices: metadata.notices.map((notice) => ({ ...notice })) } : {});
    } catch { /* Observers cannot fail or retry a completed request. */ }
  }
}

File: sdk/rust/src/response_metadata.rs
//! Optional advisory metadata, independent of API success and error payloads.

use serde::{Deserialize, Serialize};
use std::sync::{Arc, Mutex};

#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct ResponseNotice {
    pub code: String,
    pub level: String,
    pub message: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ResponseMetadata {
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_notices"
    )]
    pub notices: Option<Vec<ResponseNotice>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct APIResponse<T> {
    pub data: T,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        deserialize_with = "deserialize_notices"
    )]
    pub notices: Option<Vec<ResponseNotice>>,
}

pub type ResponseObserver = Arc<dyn Fn(&ResponseMetadata) + Send + Sync>;

pub(crate) fn notices(value: &serde_json::Value) -> Option<Vec<ResponseNotice>> {
    let mut result = Vec::new();
    for item in value.as_array()? {
        if let Ok(notice) = serde_json::from_value::<ResponseNotice>(item.clone()) {
            if !result.contains(&notice) {
                result.push(notice);
            }
        }
    }
    (!result.is_empty()).then_some(result)
}

pub(crate) fn deserialize_notices<'de, D: serde::Deserializer<'de>>(
    d: D,
) -> Result<Option<Vec<ResponseNotice>>, D::Error> {
    Ok(notices(&serde_json::Value::deserialize(d)?))
}

#[derive(Clone, Default)]
pub(crate) struct ResponseContext {
    pub observer: Option<ResponseObserver>,
    pub collector: Option<Arc<Mutex<Vec<ResponseNotice>>>>,
}

impl std::fmt::Debug for ResponseContext {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ResponseContext").finish_non_exhaustive()
    }
}

impl ResponseContext {
    pub fn observe(&self, metadata: ResponseMetadata) {
        if let (Some(collector), Some(notices)) = (&self.collector, &metadata.notices) {
            let mut collected = collector.lock().unwrap_or_else(|e| e.into_inner());
            for notice in notices {
                if !collected.contains(notice) {
                    collected.push(notice.clone());
                }
            }
        }
        if let Some(observer) = &self.observer {
            // A callback panic must never turn a completed write into a retryable failure.
            let _ = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| observer(&metadata)));
        }
    }
}

/// Only these response contracts declare top-level notice metadata.
pub(crate) fn declares_body_metadata(url: &str, method: &reqwest::Method) -> bool {
    let Ok(url) = url::Url::parse(url) else {
        return false;
    };
    let path: Vec<_> = url.path().trim_matches('/').split('/').collect();
    if matches!(path.as_slice(), ["api", "v1", "identities", _, "avatar"]) {
        return method == reqwest::Method::PUT;
    }
    matches!(
        path.as_slice(),
        ["api", "v1", "identities"]
            | ["api", "v1", "identities", _]
            | ["api", "v1", "identities", _, "companion"]
            | [
                "api",
                "v1",
                "identities",
                _,
                "companion",
                "activations",
                _,
                "messages"
            ]
            | [
                "api",
                "v1",
                "identities",
                _,
                "contacts",
                _,
                "access" | "permissions"
            ]
            | [
                "api",
                "v1",
                "identities",
                _,
                "contact-permissions" | "contact-communication-policies"
            ]
            | [
                "api",
                "v1",
                "contacts",
                _,
                "communication-policy" | "communication-preview"
            ]
            | ["api", "v1", "mail", "mailboxes", _]
            | ["api", "v1", "phone", "numbers", _]
    )
}
